### PR TITLE
Update nixpkgs to 21.05

### PR DESCRIPTION
**Pull request by marcin-rzeznicki**
Building `hoff.img` seems to work locally

### CODE REVIEW

**Comment by marcin-rzeznicki** [comment]
Fails on CI with
`./build-package.sh: line 20: 28565 Segmentation fault      mkdir -p "$PKGNAME/DEBIAN"`

Not enough memory?

**Comment by bertptrs** [comment]
Likely a weird interaction between the system C library and the Nix C library. You should be able to fix this by setting `LD_LIBRARY_PATH` to the appropriate directory.

**Comment by fatho** [comment]
> Likely a weird interaction between the system C library and the Nix C library. You should be able to fix this by setting `LD_LIBRARY_PATH` to the appropriate directory.

With the "appropriate directory" being nothing at all. In our other CI scripts, I found

```yaml
      # For some reason, Semaphore CI sets this variable, but it causes the nix installation to fail
      - unset LD_LIBRARY_PATH
```

**Comment by marcin-rzeznicki** [comment]
Still the same
`./build-package.sh: line 20: 28533 Segmentation fault      mkdir -p "$PKGNAME/DEBIAN"`

**Comment by fatho** [comment]
Hmm. I see two options
- trying to debug the job on CI (the "SSH Debug" button in the semaphore interface)
- getting rid of fakeroot (not sure if that's the cause, just a hunch because it does some unorthodox stuff), because IIRC newer `dpkg` versions allow overriding the permissions of the files it packages without pretending that the files are owned by root on disk.

**Comment by marcin-rzeznicki** [comment]
I'll try that - but then upgrading OS on Sempahore from `18.04` could help

**Comment by marcin-rzeznicki** [comment]
OK, `fakeroot` was the problem. Thx @fatho

File: nixpkgs-pinned.nix [hunk 1]
```diff
@@ -1,4 +1,4 @@
 fetchTarball {
-  url = "https://github.com/NixOS/nixpkgs/archive/b0bbacb52134a7e731e549f4c0a7a2a39ca6b481.tar.gz";
-  sha256 = "15ix4spjpdm6wni28camzjsmhz0gzk3cxhpsk035952plwdxhb67";
+  url = "https://github.com/NixOS/nixpkgs/archive/7e9b0dff974c89e070da1ad85713ff3c20b0ca97.tar.gz";
+  sha256 = "1ckzhh24mgz6jd1xhfgx0i9mijk6xjqxwsshnvq789xsavrmsc36";
```
Comment (bertptrs) [inline nixpkgs-pinned.nix:3]: I know this is scope creep but could you implement `niv` so that the next time this needs to happen it's a brain-dead task?

**Review by bertptrs — APPROVED** [review]
Changes LGTM, left one possible improvement.

File: nixpkgs-pinned.nix (hunk 1 above)
Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix:3]: Yeah, @fatho wants to get rid of `fakeroot` so I can actually work on the issue a bit more - why not

Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix]: Done in d6e5eba

Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix]: But I will ask the expert: @FPtje could you take a look at d6e5eba - I added `niv` to `hoff`

File: nixpkgs-pinned.nix [hunk 2]
```diff
@@ -1,4 +1,8 @@
-fetchTarball {
-  url = "https://github.com/NixOS/nixpkgs/archive/b0bbacb52134a7e731e549f4c0a7a2a39ca6b481.tar.gz";
-  sha256 = "15ix4spjpdm6wni28camzjsmhz0gzk3cxhpsk035952plwdxhb67";
-}
+let
```
Comment (FPtje) [inline nixpkgs-pinned.nix:1]: Could you turn this nix file into a function? I would expect `import ./nixpkgs-pinned.nix` to return a lambda. You can add arguments to make the `config` and `overlays` configurable:

```nix
{ overlays ? [] # additional overlays
, config ? {} # Imported configuration
}:
```

The `config` can then be imported as such:

```nix
config = {
      imports = [ config ];
};
```

**Review by FPtje — APPROVED** [review]
LGTM aside from nit!

File: nixpkgs-pinned.nix (hunk 2 above)
Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix:1]: Thanks for a suggestion - what is it for?

**Comment by tomwassenberg** [comment]
> I'll try that - but then upgrading OS on Sempahore from 18.04 could help
>
> EDIT: Or no - they do not have anything more recent facepalm

Might not be that relevant any more, but Semaphore did add a 20.04 image [a few weeks ago](https://docs.semaphoreci.com/reference/semaphore-changelog/#week-of-may-31-2021) :+1:

**Comment by marcin-rzeznicki** [comment]
> > I'll try that - but then upgrading OS on Sempahore from 18.04 could help
> > EDIT: Or no - they do not have anything more recent facepalm
> 
> Might not be that relevant any more, but Semaphore did add a 20.04 image [a few weeks ago](https://docs.semaphoreci.com/reference/semaphore-changelog/#week-of-may-31-2021) +1

Thanks! I simply checked the Semaphore docs for possible values. That's also a good solution, but in fact removing the problematic `fakeroot` had the nice side-effect of simplifying the scripts, so let's stick with that.

File: nixpkgs-pinned.nix (hunk 2 above)
Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix:1]: Oh, one more thing. I figured that you need this for configurability @FPtje . Please note that the whole nix stuff is used here only for packaging the app. Normal development does not warrant Nix.

Comment (FPtje) [reply nixpkgs-pinned.nix:1]: It is mostly for consistency with our other repositories. The expectation is that an import of a nixpkgs file is a function. The extra configuration is not currently needed.

Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix:1]: Makes sense - will fix

Comment (marcin-rzeznicki) [reply nixpkgs-pinned.nix:4]: Fixed in 155953e

**Comment by marcin-rzeznicki** [comment]
Thanks for reviews

@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @marcin-rzeznicki, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as 60bbb519099ba91b37eaf52c8ea997e7992ec457, waiting for CI …

**Pull request merged**